### PR TITLE
Closes #1300 - Improve Performance of DataFrame Display

### LINKED ISSUE #1300: Improve performance of dataframe display
Interactive users working in Jupyter notebooks frequently print out the first and last few rows of a dataframe to see sample values or check the results of a transformation. The syntax for this is just to execute `df` as a statement, which shows an html table of the head and tail of the dataframe, just like pandas does. With pandas, this operation is instantaneous. I believe it is also fast with single-node arkouda, but when running with many nodes, it can take 5-10 seconds to show this "quick" view of the dataframe. As a developer, this feels like an annoyingly trivial place to worry about performance, but for the user, viewing a dataframe is probably the operation with the widest gap between expected and actual runtime, and it is so common that it tends to bog down interactive workflows. As a goal, I'd love to get to the point where a 20-column dataframe can display in under 1 second.

Probably the first thing to do is create a benchmark that builds a 10_000 x 20 dataframe of mixed dtypes (including Categoricals, which are especially slow) and simulates displaying it by calling `df._repr_html_()`. That call is the one to measure for the benchmark. I'm hoping that a 16 node CS run of this benchmark would replicate the several seconds of latency we're experiencing on our typical node allocations.

In theory, we should be able to make this operation go quite fast. The amount of data transferred to the client and displayed by this operation is bounded by the dimensions of the dataframe and by the pandas options `'display.max_rows'` and `'display.max_columns'`, which default to something under 100 and in practice never get set above a few thousand. For our test dataframe with 20 columns and a `display.max_rows=100`, that means 2k elements and ~16KB (depending on strings) transferred, so a very small amount of data. The problem today is that the client-side code for `df._repr_html_()` loops over columns and slices the head and tail of each column as a separate indexing message, which incurs both message overhead and comm aggregation startup for each column. I think we could address this by making a dedicated server message that processes all columns within a single message and bypasses aggregation for this small amount of data.

### PR DESCRIPTION
**Pull request by Ethan-DeBandi99**
Closes #1300 

- Updates the `_get_head_tail()` method to function by creating a single server message instead of 1 per column in the DataFrame. This was done by adding `_get_head_tail_server()` which allows us to benchmark the new code against the old. The old function has been retained for benchmarking.
- Adds a `DataFrameIndexingMsg.chpl` to parse and process the server message for indexing the columns. The message is configured so the the column type, column name (in the df), column object name (server-side) are sent to the server to indexing to the appropriate head/tail. 
- Benchmarking configured to test `_rep_html()` which now calls `_get_head_tail_server()`. The benchmark also checks `_get_head_tail_server()` and `_get_head_tail()` directly to allow for easy comparison. 

Per the issue description, @joshmarshall1 and I did this without using aggregation because the datasets should be relatively small.

Benchmarking on a single node shows an performance already, but we will need to benchmark this on a multi-node system. Results from single node:

```bash
array size = 10,000
number of trials =  5
>>> arkouda dataframe display
numLocales = 1, N = 10,000
  _repr_html_ Average time = 0.0401 sec
  _repr_html_ Average rate = 0.03 GiB/sec
  _get_head_tail_server Average time = 0.0406 sec
  _get_head_tail_server Average rate = 0.03 GiB/sec
  _get_head_tail Average time = 0.0825 sec
  _get_head_tail Average rate = 0.01 GiB/sec

```

@reuster986 - I will leave it up to you if you would like to request an out of band benchmark.

### CODE REVIEW

**Comment by Ethan-DeBandi99** [comment]
Added @joshmarshall1 as a reviewer even though he helped write this code. There are some elements that I handled that would be good for him to review as well.

**Review by joshmarshall1 — COMMENTED** [review]
As a participant in developing this, I won't approve it, but after a second review I think we should go through and add comments throughout the chpl code, since there are almost none, just to make future maintenance easier. I also noticed a few places that TODO tags were left in and a stray #

File: benchmarks/dataframe.py
Comment (reuster986) [inline benchmarks/dataframe.py:20]: I'm thinking we should cycle deterministically through the types, rather than choosing them randomly. If today's run gets 10 Categorical columns and tomorrow's run gets 2, there will be a big difference in runtime that doesn't reflect anything about the code being benchmarked. Generating random *data* of a particular type is fine, though.

File: benchmarks/dataframe.py
Comment (reuster986) [inline benchmarks/dataframe.py:22]: This is fine for this benchmark's default problem size of 10**4, but just for future reference, the scalable way to generate a N-long ak.Strings array of 5-character strings is `ak.random_strings_uniform(minlen=5, maxlen=6, size=N)`.

File: benchmarks/dataframe.py
Comment (reuster986) [inline benchmarks/dataframe.py:27]: `ak.random_strings_uniform()` would be useful here too.

File: arkouda/dataframe.py
Comment (reuster986) [inline arkouda/dataframe.py:444]: I really love that you benchmarked both the old and new versions of this function! Although the single-node results look really promising, I'm wondering if we should look at how those benchmarks compare on the 16-node cases before we switch the implementation. We could merge this PR while still using `self._get_head_tail()`, and then if the benchmarks look good, it would be a one-line PR to change the implementation. On the other hand, it would save us some trouble if there's a hidden gotcha in the multi-locale case.

**Review by reuster986 — CHANGES_REQUESTED** [review]
Looks good overall! Just a couple requested changes in the benchmark, and a suggestion about when/how we switch to the new implementation.

File: arkouda/dataframe.py
Comment (Ethan-DeBandi99) [reply arkouda/dataframe.py]: Removed

File: src/DataFrameIndexingMsg.chpl
Comment (Ethan-DeBandi99) [reply src/DataFrameIndexingMsg.chpl]: Removed

File: src/DataFrameIndexingMsg.chpl
Comment (Ethan-DeBandi99) [reply src/DataFrameIndexingMsg.chpl]: removed

File: benchmarks/dataframe.py
Comment (Ethan-DeBandi99) [reply benchmarks/dataframe.py]: I updated this to use `x%4` to evenly distribute the column types.

File: benchmarks/dataframe.py
Comment (Ethan-DeBandi99) [reply benchmarks/dataframe.py]: I did not know that we had this. I updated to use `ak.random_strings_uniform`.

File: benchmarks/dataframe.py
Comment (Ethan-DeBandi99) [reply benchmarks/dataframe.py]: Updated to use `ak.random_strings_uniform`

File: arkouda/dataframe.py
Comment (Ethan-DeBandi99) [reply arkouda/dataframe.py:444]: This makes sense to me. I update `_repr_html_()` and `__repr__()` to use `self._get_head_tail()`. The benchmark has been updated to remove `_repr_html_()` from the operations to test and will now only directly compare `_get_head_tail()` and `get_head_tail_server()`

**Comment by Ethan-DeBandi99** [comment]
Updated benchmark results on single node with requested updates from @reuster986 implemented.

```bash
array size = 10,000
number of trials =  5
>>> arkouda dataframe display
numLocales = 1, N = 10,000
  _get_head_tail_server Average time = 0.0261 sec
  _get_head_tail_server Average rate = 0.05 GiB/sec
  _get_head_tail Average time = 0.0547 sec
  _get_head_tail Average rate = 0.02 GiB/sec
```

**Review by reuster986 — APPROVED** [review]
Looks great, thanks!

File: benchmarks/dataframe.py
Comment (pierce314159) [inline benchmarks/dataframe.py:79]: The description needs updating

File: benchmarks/dataframe.py
Comment (pierce314159) [inline benchmarks/dataframe.py:82]: I'm not sure but `--size` might need updating as well

File: src/DataFrameIndexingMsg.chpl
Comment (pierce314159) [inline src/DataFrameIndexingMsg.chpl:16]: do we need this import if we're not using aggregation?

File: src/DataFrameIndexingMsg.chpl
Comment (pierce314159) [inline src/DataFrameIndexingMsg.chpl:182]: does segarray also support `uint` as possible type for values?

**Review by pierce314159 — COMMENTED** [review]
Couple comments, nothing major. The only thing that might need to be updated is "support uint for segarray values" question? But the logic all looks good to me

File: benchmarks/dataframe.py
Comment (Ethan-DeBandi99) [reply benchmarks/dataframe.py:60]: To be completely honest, I copied this out of another benchmark and did not update it. I will take a look at argsort and update accordingly.

File: src/DataFrameIndexingMsg.chpl
Comment (Ethan-DeBandi99) [reply src/DataFrameIndexingMsg.chpl:182]: updated to properly handle supported types.

File: arkouda/dataframe.py
Comment (Ethan-DeBandi99) [reply arkouda/dataframe.py:464]: Added comment.

File: benchmarks/dataframe.py
Comment (Ethan-DeBandi99) [reply benchmarks/dataframe.py:60]: We talked offline about this, but updated this to address nbytes properly.

File: benchmarks/dataframe.py
Comment (Ethan-DeBandi99) [reply benchmarks/dataframe.py:79]: Updated

File: benchmarks/dataframe.py
Comment (Ethan-DeBandi99) [reply benchmarks/dataframe.py:82]: Updated the description.

File: src/DataFrameIndexingMsg.chpl
Comment (Ethan-DeBandi99) [reply src/DataFrameIndexingMsg.chpl:16]: Good point. This has been removed.

File: src/DataFrameIndexingMsg.chpl
Comment (Ethan-DeBandi99) [reply src/DataFrameIndexingMsg.chpl:29]: Updated.

**Review by pierce314159 — APPROVED** [review]
Looks good!

**Comment by Ethan-DeBandi99** [comment]
@reuster986 not going to lie, I am a bit confused as to why a list can't index into rows. Can you give me some details on your thought process here? I will get it pulled out though

**Pull request merged**
